scripts/create-client-config.sh

```diff
@@ -0,0 +1,138 @@
+# creates a secret with the contents of the kubeconfig file. Lastly, the script creates a
+# ClientObject that references the secret.
+#
+# This script requires the following to be installed:
+#
+#    - kubectl
+#    - yq
+#
+# TODO Accept multiple values for the src-context option and generate a kubeconfig with
+#      entries for each
+
+set -e
+
+OPTS=$(getopt -o h --long src-context:,src-kubeconfig:,dest-context:,dest-kubeconfig:,namespace:,serviceaccount,output-dir:,help -n 'create-client-config' -- "$@")
+
+eval set -- "$OPTS"
+
+function help() {
+    echo
+    echo "Syntax: create-client-config.sh [options]"
+    echo "Options:"
+    echo "src-context     The context for the source cluster that contains the service account. This or the src-kubeconfig option must be set."
+    echo "src-kubeconfig   The kubeconfig for the source cluster that contains the service account. This or the src-context option must be set."
+    echo "dest-context    The context for the cluster where the ClientConfig will be created."
+    echo "dest-kubeconfig  The kubeconfig for the cluster where the ClientConfig will be created."
+    echo "namespace       The namespace in which the service account exists and where the ClientConfig will be created."
+    echo "serviceaccount  The name of the service account from which the ClientConfig will be created. Defaults to k8ssandra-operator."
+    echo "output-dir      The directory where generated artifacts are written. If not specified a temp directory is created."
+}
+
+src_context=""
+src_kubeconfig=""
+dest_context=""
+dest_kubeconfig=""
+service_account="k8ssandra-operator"
+namespace=""
+output_dir=""
+
+while true; do
+  case "$1" in
+    --src-context ) src_context="$2"; shift 2 ;;
+    --src_kubeconfig ) src_kubeconfig="$2"; shift 2 ;;
+    --dest-context ) dest_context="$2"; shift 2 ;;
+    --dest-kubeconfig ) dest_kubeconfig="$2"; shift 2 ;;
+    --namespace ) namespace="$2"; shift 2 ;;
+    --serviceaccount ) service_account="$2"; shift 2 ;;
+    --output-dir ) output_dir="$2"; shift 2 ;;
+    -h | --help ) help; exit;;
+    -- ) shift; break ;;
+    * ) break ;;
+  esac
+done
+
+if [ -z "$src_context" ] && [ -z "$src_kubeconfig" ]; then
+  echo "At least one of the --src-context or --src-kubeconfig options must be specified"
+  exit 1
+fi
+
+src_context_opt=""
+src_kubeconfig_opt=""
+namespace_opt=""
+
+if [ -z "$src_context" ]; then
+  src_context=$(kubectl $src_kubeconfig config current-context)
+else
+  src_kubeconfig=""
+  src_context_opt="--context $src_context"
+fi
+
+if [ ! -z "$src_kubeconfig" ]; then
+  src_kubeconfig_opt="--kubeconfig $src_kubeconfig"
+fi
+
+if [ -z "$output_dir" ]; then
+  output_dir=$(mktemp -d)
+else
+  mkdir -p $output_dir
+fi
+
+if [ ! -z "$namespace"]; then
+  namespace_opt="-n $namespace"
+fi
+
+sa_secret=$(kubectl $src_kubeconfig_opt $src_context_opt $namespace_opt get serviceaccount $service_account -o jsonpath='{.secrets[0].name}')
```

Comment:
Is this procedure supposed to work on any cluster? On my local kind cluster, the service account exists, but has no secrets:

```
kubectl get sa k8ssandra-operator -o json
{
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"ServiceAccount\",\"metadata\":{\"annotations\":{},\"name\":\"k8ssandra-operator\",\"namespace\":\"default\"}}\n"
        },
        "creationTimestamp": "2021-08-20T15:37:14Z",
        "name": "k8ssandra-operator",
        "namespace": "default",
        "resourceVersion": "34052",
        "uid": "982796e5-0a00-4d7d-8a1c-8dd9067d9a13"
    }
}
```

Reply:
Good question. I assumed it would but I only tested against GKE so far. My understanding from the [docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/) is that the secret is automatically added to the service account. I think that behavior is controlled by an auth plugin. Maybe it is turned off for kind. I'll investigate.

Reply:
@adutra I just deployed the operator in my kind cluster as follows:

`kustomize build config/default | k apply -f -`

And here is the service account:

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{},"name":"k8ssandra-operator","namespace":"default"}}
  creationTimestamp: "2021-08-20T16:19:01Z"
  name: k8ssandra-operator
  namespace: default
  resourceVersion: "28259"
  uid: a9ab7ac2-6405-4d13-b824-919589d3c170
secrets:
- name: k8ssandra-operator-token-snqp9
```
I am curious why we are seeing different behavior.

Reply:
I did the same. But I just deleted the cluster and re-created everything from scratch, and now the service account has a token. Strange. But let's disregard this incident for now.

Reply:
I confirmed this morning that, once the syntax error line 85 is fixed, the script works and the generated secret is OK.